### ISSUE TITLE
Catching Timeout error and KeyboardInterrupt

### ISSUE THREAD

**Issue by rgildein** [canonical/charmed-openstack-upgrader#103]
If control+c is pressed the result should not be showing any traceback.  If it's not possible to connect to model, the error should provide more user friendly output. 

Example with `TimeoutError`:
```bash
x1:➜  charmed-openstack-upgrader git:(tmp/testing) ✗ cou plan --model openstack
Configuring logging... ✖
2023-10-05 22:05:48 [ERROR] function _connect timed out after 60s                                                                  [1m5.76s]
```

Example with `KeyboardInterrupt`:
```bash
x1:➜  charmed-openstack-upgrader git:(tmp/testing) ✗ cou plan --model openstack 
^CTraceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 1022, in _find_and_load
KeyboardInterrupt
thread '<unnamed>' panicked at 'Python API call failed', /github/home/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.18.3/src/err/mod.rs:790:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "/snap/charmed-openstack-upgrader/x1/bin/cou", line 5, in <module>
    from cou import main
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cou/__init__.py", line 15, in <module>
    from cou.__main__ import main
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cou/__main__.py", line 18, in <module>
    from cou import cli
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cou/cli.py", line 25, in <module>
    from cou.exceptions import COUException
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cou/exceptions.py", line 17, in <module>
    from juju.action import Action
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/juju/action.py", line 4, in <module>
    from . import model
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/juju/model.py", line 25, in <module>
    from . import provisioner, tag, utils, jasyncio
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/juju/provisioner.py", line 11, in <module>
    import paramiko
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/paramiko/__init__.py", line 22, in <module>
    from paramiko.transport import SecurityOptions, Transport
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/paramiko/transport.py", line 34, in <module>
    from cryptography.hazmat.primitives.ciphers import algorithms, Cipher, modes
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cryptography/hazmat/primitives/ciphers/__init__.py", line 11, in <module>
    from cryptography.hazmat.primitives.ciphers.base import (
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 10, in <module>
    from cryptography.exceptions import (
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cryptography/exceptions.py", line 9, in <module>
    from cryptography.hazmat.bindings._rust import exceptions as rust_exceptions
pyo3_runtime.PanicException: Python API call failed                                                                                  [0.76s]
x1:➜  charmed-openstack-upgrader git:(tmp/testing) ✗ cou plan --model openstack 
Traceback (most recent call last):
  File "/snap/charmed-openstack-upgrader/x1/bin/cou", line 8, in <module>
    sys.exit(main())
  File "/snap/charmed-openstack-upgrader/x1/lib/python3.10/site-packages/cou/__main__.py", line 23, in main
    asyncio.run(cli.entrypoint())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 1871, in _run_once
    event_list = self._selector.select(timeout)
  File "/usr/lib/python3.10/selectors.py", line 469, in select
    fd_event_list = self._selector.poll(timeout, max_ev)
KeyboardInterrupt
                                                                                                                                     [2.76s]
```

**Comment by rgildein** [comment]
During testing my fix, I found that we are not catching some Exception properly. The result is cou telling me following error if I disconnect from VPN.
```bash
(.venv) x1:➜  charmed-openstack-upgrader git:(cli/fix-error-handling) ✗ .venv/bin/cou plan --model openstack 
Analyzing cloud... \2023-10-05 22:39:57 [ERROR] Unexpected error occurred
2023-10-05 22:39:57 [ERROR] sent 1011 (unexpected error) keepalive ping timeout; no close frame received                                                                                                                                                              [1m27.23s]
```

**Comment by aieri** [comment]
Duplicates #80

**Comment by aieri** [comment]
I think a SIGINT should cause cou to terminate at the earliest possible step, which is probably between steps (interrupting within a step may leave the cloud in a non-resumable state)

**Comment by aieri** [comment]
> @aieri I proposed fix #109, where I'm not fully happy with having two try catch. One in `cou.__main__.main` and second in `cou.cli.entrypoint`. After I pushed I saw your comment and realized that you are right we should not enable to do keyboard interruption after `run_upgrade` function was executed. Is that what you mean?

Not exactly. My point is that we should balance the desire for CTRL+C to interrupt the process immediately, with avoiding to leave the cloud in a broken state. Ideally, it should always be possible to rerun cou and resume an upgrade that was interrupted with CTRL+C. I suspect that a safe way to do that would be to let whatever `UpgradeStep` is running complete, but not progress any further.
In any case, it is important that the user has a way to stop an in-progress update, at least when running non-interactively.
